### ISSUE TITLE
Golang "holds" global variable while gorilla "pump" thread alive

### ISSUE THREAD

**Issue by theromis** [gorilla/websocket#630]
My problem `go-pg` reconnect.
https://github.com/go-pg/pg/issues/1706
Backend reconnects but only after 1-2 minutes. Without active websocket connection it reconnects on second request.

**Versions**
```
 ~ go version
go version go1.15 linux/amd64
```
package version: `github.com/gorilla/websocket@v1.4.2`

**Steps to Reproduce**
Restart postgresql with websocket connection established,  `go-pg` not catching `connection lost` until websocket timeout

**Expected behavior**
Reconnection success.

**Code Snippets**
```go
const (
    // Time allowed to write a message to the peer.
    writeWait = 10 * time.Second

    // Time allowed to read the next pong message from the peer.
    pongWait = 60 * time.Second

    // Send pings to peer with this period. Must be less than pongWait.
    pingPeriod = (pongWait * 9) / 10

    // Maximum message size allowed from peer.
    maxMessageSize = 512000
)

type connection struct {
    // The websocket connection.
    ws *websocket.Conn

    // Buffered channel of outbound messages.
    send chan []byte
}

func (c *connection) readPump() {
    c.ws.SetReadLimit(maxMessageSize)
    c.ws.SetReadDeadline(time.Now().Add(pongWait))
    c.ws.SetPongHandler(func(string) error { c.ws.SetReadDeadline(time.Now().Add(pongWait)); return nil })
    for {
        _, input, err := c.ws.ReadMessage()
        if err != nil {
            /*
                if websocket.IsUnexpectedCloseError(err, websocket.CloseGoingAway) {
                    logrus.Errorf("UnexpectedCloseError: %v", err)
                }
                //logrus.Errorf("ReadMessage: %v", err)
            */
            break
        }
...
}

func serveWs(w http.ResponseWriter, r *http.Request) {
    ws, err := upgrader.Upgrade(w, r, nil)
    if err != nil {
        logrus.Errorf("upgrader: %v", err)
        return
    }
    c := &connection{
        send:      make(chan []byte, 256),
        ws:        ws,
    }
    go c.writePump()
    c.readPump()
}
```

**Comment by TaylorRex** [comment]
Which variable is Golang holding and what does it mean to "hold" the variable?   How is postgresql related to the program?  Is the client postgresql?

**Comment by theromis** [comment]
@TaylorRex 
Which variable is Golang holding and what does it mean to "hold" the variable?
Global variable/context, in my case `go-pg` context.

How is postgresql related to the problem?
`go-pg` not reconnecting.

Is the client postgresql?
Yes, client is `go-pg`

**Comment by TaylorRex** [comment]
I cannot find a reference to websockets in the go-pg code.  The websocket package does not directly or indirectly import go-pg.  There are not references to go-pg in your posted code.   Explain how go-pg is related to the problem.

What is the specific global variable in go-pg that Golang is "holding".

**Comment by theromis** [comment]
@TaylorRex sorry for not clear problem description.
Gorilla not uses postgres, my project uses `gorilla` and `go-pg` and after postgresql server restart if I have active gorilla websocket connections existsing postgresql client in my backend app continue getting connection error, and "reconnecting" only after gorilla websocket readPump exits.
postgresql connection context declared as global variable

**Comment by TaylorRex** [comment]
Does your application use go-pg in the code elided from conn.readPump? 

Show the code where the postgresql connection context is declared as a global variable.

**Comment by theromis** [comment]
I did a test and removed any db requests from read/write Pump but it didn't change behavior.

```go
var (
    DbWrite    *pg.DB
    dbPtr      = 0
    dbReadPool []*pg.DB
 )
func DbInit() (*pg.DB, []*pg.DB, error) {
........
    writeOptions, err := pg.ParseURL(writeUrl)
    if err != nil {
        return nil, nil, err
    }

    writeConn := pg.Connect(writeOptions)

    var readPool []*pg.DB
    for _, readUrl := range backends {
        readOptions, err := pg.ParseURL(readUrl)
        if err != nil {
            return nil, nil, err
        }
        readConn := pg.Connect(readOptions)
        readPool = append(readPool, readConn)
    }
    return writeConn, readPool, nil
}

func Open() error {
    var err error
    EmojifyInit()

    DbWrite, dbReadPool, err = DbInit()
    return err
}

func Close() {
    DbWrite.Close()
    for _, readDb := range dbReadPool {
        readDb.Close()
    }
}
```
Note: `pg.Connect` didn't do an actual connection, it just prepares a DB attributes, an actual connection (reconnection in case of error) happens db request `Insert/Update/...`

**Comment by TaylorRex** [comment]
> I did a test and removed any db requests from read/write Pump

It sounds like the elided code is relevant. Show the actual code for readPump, writePump and serveWs. 

Show the code for the global variable that Golang is "holding".

**Comment by theromis** [comment]
```go
    DbWrite, dbReadPool, err = DbInit()
```
this is what Golang is "holding".

```go
// write writes a message with the given message type and payload.
func (c *connection) write(mt int, payload []byte) error {
    c.ws.SetWriteDeadline(time.Now().Add(writeWait))
    return c.ws.WriteMessage(mt, payload)
}

// writePump pumps messages from the hub to the websocket connection.
func (c *connection) writePump() {
    ticker := time.NewTicker(pingPeriod)
    defer func() {
        ticker.Stop()
        logrus.Debug("writePump cleanup")
    }()
    for {
        select {
        case message, ok := <-c.send:
            if !ok {
                c.write(websocket.CloseMessage, []byte{})
                logrus.Debug("writePump CloseMessage")
                return
            }
            if err := c.write(websocket.TextMessage, message); err != nil {
                logrus.Debugf("writePump websocket.TextMessage: %v", err)
                return
            }
        case <-ticker.C:
            if err := c.write(websocket.PingMessage, []byte{}); err != nil {
                logrus.Debugf("writePump ticker.C: %v", err)
                return
            }
        }
    }
}
func (c *connection) readPump() {
    h.connect <- c
    defer func() {
        h.cleanup <- c
        logrus.Debug("readPump cleanup")
    }()
    c.ws.SetReadLimit(maxMessageSize)
    c.ws.SetReadDeadline(time.Now().Add(pongWait))
    c.ws.SetPongHandler(func(string) error { c.ws.SetReadDeadline(time.Now().Add(pongWait)); return nil })
    for {
        _, input, err := c.ws.ReadMessage()
        if err != nil {
            /*
                if websocket.IsUnexpectedCloseError(err, websocket.CloseGoingAway) {
                    logrus.Errorf("UnexpectedCloseError: %v", err)
                }
                //logrus.Errorf("ReadMessage: %v", err)
            */
            break
        }

        action := serializer.BackendAction{}
        if err := json.Unmarshal(input, &action); err != nil {
            logrus.Errorf("JSON: %v", err)
            continue
        }
        logrus.Debugf("action: %v %v", action.Action, action.Payload)
        rq := &request{c, &action}

        switch action.Action {
        case "authorize":
            if jwt, ok := action.Payload.(string); ok {
                err := c.authorizeJwt(jwt)
                resp := serializer.BackendAction{}
....
        default:
            logrus.Errorf("no mapping for: %v", action.Action)
        }
    }
}
```

**Comment by TaylorRex** [comment]
Go "holds" the package-level variables `DbWrite` and `dbReadPool` for the lifetime of the application.  The websocket package does not have any influence over the lifetime of these or any other package-level variables.

Does the code that receives on `h.cleanup` touch the database?  Does `serveWS` touch the database?  Otherwise, I don't see any relationship  between read and write pumps in the issue and the go-pg package.   I don't see how the websocket package is involved.

**Comment by theromis** [comment]
sorry for delayed answer
serveWs only websocket initializer:
```go
func serveWs(w http.ResponseWriter, r *http.Request) {
    ws, err := upgrader.Upgrade(w, r, nil)
    if err != nil {
        logrus.Errorf("upgrader: %v", err)
        return
    }
    c := &connection{
        send:      make(chan []byte, 256),
        ws:        ws,
    }
    go c.writePump()
    c.readPump()
}
```
connect/cleanup keydb(redis) related funcs
```go
func (h *hub) run() {
    keepAliveTicker := time.NewTicker(60 * time.Second)
    for {
        select {
        case conn := <-h.connect:
            // Register global connection list
            h.connections[conn] = true
        case conn := <-h.cleanup:
            // Delete from chatrooms connection list
            for chatroom, connections := range h.chatrooms {
                if _, ok := connections[conn]; ok {
                    s := &subscription{conn, chatroom, "", false}
                    r.user_deregister <- s
                    delete(connections, conn)
                    if len(connections) == 0 {
                        r.unsubscribe <- chatroom
                        delete(h.chatrooms, chatroom)
                    }
                }
            }
            // Delete from global connection list
            if _, ok := h.connections[conn]; ok {
                delete(h.connections, conn)
            }
            conn.ws.Close()
```
seems like I don't have a choice other then write a small app with websocket to catch this feature :(

**Comment by TaylorRex** [comment]
In the code posted so far,  the websocket related code does not touch the database.  

 Do you have "middleware" that wraps `serveWs` and does that middleware access the database?  Changing the call to `c.readPump()` to `go c.readPump()` might reveal issues related to middleware.

There's no code to ensure that the read pump promptly returns when the write pump returns.  I would add deferred calls to `c.ws.Close()` to the read pump and write pump and remove the call to `Close()`  from the hub.

There's no evidence that the webocket package is responsible for the issue.

**Comment by TaylorRex** [comment]
@theromis  There's not enough information to identify a bug with with the websocket package. There may be a problem with the lifetime of a goroutine, but that's outside the scope the websocket package.  This issue should be closed unless you can provide additional debugging information.

**Comment by theromis** [comment]
@TaylorRex thank you for a lot of details, unfortunately I have bunch of other tasks :( Closing for now, definitely will dig deeper once will find more details will share it here (probably reopen the issue). Closing for now.
